### ISSUE TITLE
TextArea scrolls content out of its own boundary

### ISSUE THREAD

**Issue by t-arn** [codenameone/CodenameOne#1444]
**Environment:**
libVersion 85 on iOS, Android

**Problem:**
Under some circumstances (see steps below), tapping into a TextArea to edit it will cause the content to be scrolled outside the boundaries of the TextArea (see following screenshots):
![screenshot_2015-04-14-06-53-17](https://cloud.githubusercontent.com/assets/8982757/7132725/48b05782-e28c-11e4-9450-5b8a486331d4.png)
![screenshot_2015-04-14-06-53-39](https://cloud.githubusercontent.com/assets/8982757/7132728/4c02f82c-e28c-11e4-8502-88bd39971fe0.png)

**Steps to reproduce:**
* Create a simple form:
```
  public void start ()
  {
    if (current != null)
    {
      current.show();
      return;
    }
    TextArea ta;
    Form f = new Form("TextArea bug sample");
    f.setLayout(new BoxLayout(BoxLayout.Y_AXIS));
    f.setScrollableY(true);
    f.addComponent(new Label("To reproduce the bug, scroll the TextArea"));
    f.addComponent(new Label("to the bottom and then tap into it to edit."));
    f.addComponent(new Label("Label 3"));
    f.addComponent(new Label("Label 4"));
    f.addComponent(new Label("Label 5"));
    f.addComponent(new Label("Label 6"));
    f.addComponent(new Label("Label 7"));
    f.addComponent(new Label("Label 8"));
    f.addComponent(new Label("Label 9"));
    f.addComponent(new Label("Now comes the TextArea:"));
    ta = new TextArea(10, 20);
    ta.setGrowLimit(10);
    // add more lines than grow limit
    ta.setText("line1\nline2\nline3\nline4\nline5\nline6\nline7\nline8\nline9\nline10\nline11\nline12\nline13\nline14");
    f.addComponent(ta);
    f.show();
  }
```
* Build and start the app
* Scroll the TextArea down to the bottom
* Tap into the TextArea

**Comment by shannah** [comment]
I'm just looking at the test case and trying to narrow down where the problem is occuring.  Here is the component inspector for the TextArea when it is scrolled down.

![image](https://cloud.githubusercontent.com/assets/2677562/7478434/a130663e-f30f-11e4-87c9-76212b4146d0.png)

I notice that absY < Y.  This is because the implementation of Component.getAbsoluteY() subtracts the scroll position of the component itself.

~~~~
    public int getAbsoluteY() {
        int y = getY() - getScrollY();
        Container parent = getParent();
        if (parent != null) {
            y += parent.getAbsoluteY();
        }
        return y;
    }
~~~~

Is this correct behaviour?  My gut tells me that absY on the TextArea should not be affected by the scroll position of the text area itself - only by the scroll of its parent.  Am I wrong, @codenameone ?

**Comment by shannah** [comment]
This is a bit ugly.  It seems that "correctly", getAbsoluteY() should not be affected by the scrollY of the component itself - only its ancestors.  E.g. the following implementation would be more correct:

~~~~
    public int getAbsoluteY() {
        int y = getY();
        Container parent = getParent();
        if (parent != null) {
            y -= parent.getScrollY();
            y += parent.getAbsoluteY();
        }
        return y;
    }
~~~~

In fact, this change fixes the TextArea issue ... BUT there are many places in code where this old calculation was counted on - and thus we have lines like:

~~~~
int absY = cmp.getAbsoluteY() + cmp.getScrollY();
~~~~

in many places.

So do we:

1. Fix the issue and try to identify places where the old behaviour was depended upon.
or
2. Continue with this definition of absoluteY and continue offsetting it.

For purity, the first option sounds nice, but it isn't feasible as it could break many existing apps not under our control.

Therefore, I suppose the best solution at this point is to add some javadocs to explain getAbsoluteX() and getAbsoluteY() - and update all of the ports individually for the TextArea.

**Comment by shannah** [comment]
For JavaSE port I couldn't seem to get the default scrolling of the JTextArea to cooperate so there is still a rough edge there.  But the overpaint is fixed.  Still may need to fix for iOS, Android, and Javascript ports.  Investigating that now.

**Comment by shannah** [comment]
The issue is now fixed in JavaSE, iOS, and Android.  Awaiting review.  Then I will merge them into the 3.0 branch and close this issue.

**Comment by shannah** [comment]
Since this won't be included in 3.0, I'm closing this.

**Comment by t-arn** [comment]
I just made a new iOS build with libVersion 88 and the bug #1444 appeared again (or still? I'm not sure I tested it successfully on iOS before). Tested on iPad 2 mini with iOS 8.3

(By the way: On Android, the bugfix works: The problem does not occur on Android anymore.)

Tom
